sql/parser/eval.go

```diff
@@ -658,27 +658,42 @@ func transformAlias(name string) string {
 }
 
 func evalFuncExpr(expr *FuncExpr) (Datum, error) {
-	name := strings.ToLower(expr.Name.String())
-	name = transformAlias(name)
-	b, ok := builtins[name]
+	name := transformAlias(strings.ToLower(expr.Name.String()))
+
+	candidates, ok := builtins[name]
 	if !ok {
-		return DNull, fmt.Errorf("%s: unknown function", expr.Name)
-	}
-	if b.nArgs != -1 && b.nArgs != len(expr.Exprs) {
-		return DNull, fmt.Errorf("%s: incorrect number of arguments: %d vs %d",
-			expr.Name, b.nArgs, len(expr.Exprs))
+		return DNull, fmt.Errorf("unknown function %s", expr.Name)
 	}
 
 	args := make(DTuple, 0, len(expr.Exprs))
+	types := make([]string, 0, len(expr.Exprs))
 	for _, e := range expr.Exprs {
 		arg, err := EvalExpr(e)
 		if err != nil {
 			return DNull, err
 		}
 		args = append(args, arg)
+		types = append(types, arg.Type())
+	}
+
+	var impl func(DTuple) (Datum, error)
+	for _, candidate := range candidates {
+		if candidate.types == nil || reflect.DeepEqual(candidate.types, types) {
```

Comment:
Not fond of the use of `reflect.DeepEqual` outside of a testing environment. I think you should add a `builtin.match(types)` which returns true if `builtin.types` is `nil` or the list of types (which it would loop over) matches the passed in types.